addons/l10n_ke_edi_oscu/models/account_move.py

```diff
@@ -0,0 +1,556 @@
+            except Exception:
+                pass
+        return super()._get_edi_decoder(file_data, new=new)
+
+    def _l10n_ke_oscu_import_invoice(self, invoice, data, is_new):
+        """ Decodes the json content from eTIMS into an Odoo move.
+
+        This method is passed as the EDI decoder in the case where the file is recognised as an OSCU
+        JSON representation of a vendor bill.
+
+        :param dictionary data: the dictionary with the content to be imported
+        :param boolean is_new:  whether the vendor bill is newly created or to be updated
+        :returns:               the imported vendor bill
+        """
+        company = self.company_id
+        content = json.loads(data['content'])
+        message_to_log = []
+
+        self.move_type = {
+            'S': 'in_invoice',
+            'R': 'in_refund',
+        }.get(content['rcptTyCd'], 'in_invoice')
+
+        branches = self.env['res.partner'].search([('vat', 'ilike', content['spplrTin'])])
+        if (branch := branches.filtered(lambda branch: branch.l10n_ke_oscu_branch_code == content['spplrBhfId'])):
+            self.partner_id = branch
+        else:
+            self.partner_id = self.env['res.partner'].create({
+                'name': content['spplrNm'],
+                'vat': content['spplrTin'],
+                'l10n_ke_oscu_branch_code': content['spplrBhfId'],
+            })
+            message_to_log.append(_(
+                "A vendor with a matching Tax ID and Branch ID was not found. "
+                "One with the corresponding details was created."
+            ))
+            message_to_log.append("")
+
+        self.invoice_date = datetime.strptime(content['salesDt'], '%Y%m%d').date()
+        self.l10n_ke_control_unit = content['spplrSdcId']
+
+        lines_dict = self.env['product.product']._l10n_ke_assign_products_from_json(
+            {item['itemSeq']: item for item in content['itemList']}
+        )
+
+        tax_rate_map = {code: content[f'taxRt{code}'] for code in TAX_CODE_LETTERS}
+        uom_codes = [line['qtyUnitCd'] for line in lines_dict.values()]
+        uom_map = {
+            unit_code.code: uom_id
+            for unit_code, uom_id in self.env['uom.uom']._read_group(
+                domain=[('l10n_ke_quantity_unit_id.code', 'in', uom_codes)],
+                groupby=['l10n_ke_quantity_unit_id'],
+                aggregates=['id:min'],
+            )
+        }
+        for item in lines_dict.values():
+            if (taxes := item.get('product') and item['product'].supplier_taxes_id.filtered(lambda tax: tax.company_id == self.company_id)):
+                for tax in taxes:
+                    if tax.l10n_ke_tax_type_id.code == item['taxTyCd'] and tax.amount == tax_rate_map[item['taxTyCd']]:
+                        item['tax'] = tax
+                        break
+
+            # If we can't select a tax from the product,
+            if not item.get('tax'):
```

Comment:
In principle, I think it's safer this way: although in the default taxes there is only one 16%, one 8%, one 0% and one VAT-exempt purchase tax, I can imagine the user might perhaps wish to create more taxes with the same percentages.

Reply:
Still seems a bit magical and hypothetical to me :thinking:  Was it validated with the PO ?

Reply:
I think the current code is overkill, as it will be one tax per type.  Right now, we check if the taxes are in the default taxes of the product first while that is not necessary: we can immediately map with the right tax.